File: myai/platform.py
"""Cross-platform OS, architecture, and bundled runtime resolution."""
from __future__ import annotations

import os
import platform
from pathlib import Path


def detect_platform() -> dict[str, str]:
    system = platform.system().lower()
    current_os = "windows" if system == "windows" else "macos" if system == "darwin" else "linux"
    arch = "arm64" if platform.machine().lower() in {"arm64", "aarch64"} else "x64"
    return {
        "os": current_os,
        "arch": arch,
        "runtime_tag": f"{'win' if current_os == 'windows' else 'darwin' if current_os == 'macos' else 'linux'}-{arch}",
        "executable": "llama-server.exe" if current_os == "windows" else "llama-server",
    }


def resolve_runtime_binary(root: str | Path, binary: str = "llama-server") -> Path:
    info = detect_platform()
    name = info["executable"] if binary == "llama-server" else binary
    return Path(root).resolve() / "runtime" / info["runtime_tag"] / name


def paths_equal(left: str | Path, right: str | Path) -> bool:
    a, b = Path(left).resolve(), Path(right).resolve()
    return os.path.normcase(a) == os.path.normcase(b)

File: src/utils/platform.ts
import os from "node:os";
import path from "node:path";

export type PlatformOS = "windows" | "macos" | "linux";
export type PlatformArch = "x64" | "arm64";

export interface PlatformInfo {
  os: PlatformOS;
  arch: PlatformArch;
  runtimeTag: string;
  executableName: string;
}

export function detectPlatform(): PlatformInfo {
  const currentOs: PlatformOS = process.platform === "win32"
    ? "windows" : process.platform === "darwin" ? "macos" : "linux";
  const arch: PlatformArch = process.arch === "arm64" ? "arm64" : "x64";
  return {
    os: currentOs,
    arch,
    runtimeTag: `${currentOs === "windows" ? "win" : currentOs === "macos" ? "darwin" : "linux"}-${arch}`,
    executableName: currentOs === "windows" ? "llama-server.exe" : "llama-server"
  };
}

export function resolveRuntimeBinary(root: string, binary = "llama-server"): string {
  const info = detectPlatform();
  const name = info.os === "windows" && binary === "llama-server" ? info.executableName : binary;
  return path.resolve(root, "runtime", info.runtimeTag, name);
}

export function normalizePath(value: string): string {
  return path.normalize(path.resolve(value));
}

export function pathsEqual(left: string, right: string): boolean {
  const a = normalizePath(left);
  const b = normalizePath(right);
  return process.platform === "linux" ? a === b : a.toLowerCase() === b.toLowerCase();
}

export const hostPlatform = os.platform();
